Dogfooding instructions for .NET 8 NativeAOT on macOS x64 via Rosetta

We're extremely happy that NativeAOT on Mac x64/arm64 should now work in .NET 8 daily builds.
As one of the early proponents of CoreRT, we're looking forward to help beta test this in a large real world application.

However, I'm running into issues with using the .NET 8 daily builds of the SDK with ILC.
We need to build macOS universal binaries (eventually) so all of our builds of x64 are done via Rosetta 2 on an Apple Silicon device.

Here's the repro:
https://github.com/christianscheuer/nativeaot-net8

Steps:
- I installed macOS x64 SDK version 8.0.100-alpha.1.22605.4 from github.com/dotnet/installer
- All commands are run from a Rosetta Terminal.
- The SDK host is x64, the SDK is x64.
- Run `./build.sh` which essentially just ensures we're in Rosetta, then runs `dotnet publish -c Release -r osx-x64`

Expected result:
- It builds

Actual result:
```
MSBuild version 17.5.0-preview-22564-01+d14b74d6f for .NET
  Determining projects to restore...
  Restored /Users/chr/d/d2/nativeaot-net8/nativeaot-net8.csproj (in 327 ms).
/usr/local/share/dotnet/x64/sdk/8.0.100-alpha.1.22605.4/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.RuntimeIdentifierInference.targets(275,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy [/Users/chr/d/d2/nativeaot-net8/nativeaot-net8.csproj]
  nativeaot-net8 -> /Users/chr/d/d2/nativeaot-net8/bin/Release/net8.0/osx-x64/nativeaot-net8.dll
/Users/chr/.nuget/packages/microsoft.dotnet.ilcompiler/8.0.0-alpha.1.22559.2/build/Microsoft.DotNet.ILCompiler.SingleEntry.targets(51,5): warning : Delete explicit 'Microsoft.DotNet.ILCompiler' package reference in your project file. Explicit 'Microsoft.DotNet.ILCompiler' package reference can run into version errors. [/Users/chr/d/d2/nativeaot-net8/nativeaot-net8.csproj]
/Users/chr/.nuget/packages/microsoft.dotnet.ilcompiler/8.0.0-alpha.1.22559.2/build/Microsoft.NETCore.Native.Publish.targets(70,5): error : Add a PackageReference for 'runtime.osx-arm64.Microsoft.DotNet.ILCompiler' to allow cross-compilation for x64 [/Users/chr/d/d2/nativeaot-net8/nativeaot-net8.csproj]
```

What's weird is that it claims it can't do cross-compilation, which I'm not actually doing. This is the x64 SDK running in an x64 terminal trying to build an x64 release. Furthermore, it requests the arm64 PackageReference. Adding that reference changes nothing, it reports the same error. It also complains about an explicit package reference that I'm not seeing.

I tried following the instructions on dotnet/installer to add the Nuget.config, but that ends up in yet other errors.

I shared the output of `dotnet --info` below.

Is there anything I should do differently to consume and test .NET 8 NativeAOT for macOS x64?
Are there dogfooding instructions for this scenario somewhere that I haven't located?


<details>

<summary>`dotnet --info` output:</summary>

```
.NET SDK:
 Version:   8.0.100-alpha.1.22605.4
 Commit:    af14110139

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  12.5
 OS Platform: Darwin
 RID:         osx.12-x64
 Base Path:   /usr/local/share/dotnet/x64/sdk/8.0.100-alpha.1.22605.4/

Host:
  Version:      8.0.0-alpha.1.22559.2
  Architecture: x64
  Commit:       dd7fdb723d

.NET SDKs installed:
  5.0.408 [/usr/local/share/dotnet/x64/sdk]
  6.0.403 [/usr/local/share/dotnet/x64/sdk]
  8.0.100-alpha.1.22605.4 [/usr/local/share/dotnet/x64/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 5.0.17 [/usr/local/share/dotnet/x64/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.11 [/usr/local/share/dotnet/x64/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.0-alpha.1.22558.5 [/usr/local/share/dotnet/x64/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 5.0.17 [/usr/local/share/dotnet/x64/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.11 [/usr/local/share/dotnet/x64/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.0-alpha.1.22557.12 [/usr/local/share/dotnet/x64/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.0-alpha.1.22559.2 [/usr/local/share/dotnet/x64/shared/Microsoft.NETCore.App]

Other architectures found:
  arm64 [/usr/local/share/dotnet]
    registered at [/etc/dotnet/install_location_arm64]

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
```

</details>

cc @jkotas

Thanks so much, everybody.

@am11 these instructions on consuming daily builds are extremely valuable, thank you for sharing. So much better not to pollute the regular dotnet installation with these daily builds but have more control. The nuget package clearing is neat too.

I got the x64 build working as cross compilation now with @am11's instructions now using the arm64 SDK, and it works just fine using either x64 or arm64 terminal.

I can also target arm64 using the arm64 SDK using those instructions, when in an arm64 terminal. Curiously, if I am in a Rosetta terminal and invoke the arm64 SDK, instead of getting a nice error saying I shouldn't do that, it fails with linker errors.

To repro that, follow @am11's instructions to install dotnet8.
- Open Rosetta Terminal (or type `arch -x86_64 zsh` to open a sub shell in x64 mode).
- Follow @am11's repro instructions, then run this at the publish step: ` dotnet8 publish helloworldapp -a arm64 -o helloworldapp/distarm -p:PublishAot=true`


<details>
<summary>dotnet8 publish output</summary>

```
MSBuild version 17.5.0-preview-22564-01+d14b74d6f for .NET
  Determining projects to restore...
  Restored /Users/chr/d/d2/helloworldapp/helloworldapp.csproj (in 89 ms).
/Users/chr/.dotnet8/sdk/8.0.100-alpha.1.22605.6/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.RuntimeIdentifierInference.targets(275,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy [/Users/chr/d/d2/helloworldapp/helloworldapp.csproj]
  helloworldapp -> /Users/chr/d/d2/helloworldapp/bin/Debug/net8.0/osx-arm64/helloworldapp.dll
/Users/chr/.nuget/packages/microsoft.dotnet.ilcompiler/8.0.0-alpha.1.22559.2/build/Microsoft.DotNet.ILCompiler.SingleEntry.targets(51,5): warning : Delete explicit 'Microsoft.DotNet.ILCompiler' package reference in your project file. Explicit 'Microsoft.DotNet.ILCompiler' package reference can run into version errors. [/Users/chr/d/d2/helloworldapp/helloworldapp.csproj]
  Generating native code
  ld: warning: ignoring file /Users/chr/.nuget/packages/runtime.osx-arm64.microsoft.dotnet.ilcompiler/8.0.0-alpha.1.22559.2/framework/libSystem.Net.Security.Native.a, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
  ld: warning: ignoring file /Users/chr/.nuget/packages/runtime.osx-arm64.microsoft.dotnet.ilcompiler/8.0.0-alpha.1.22559.2/sdk/libbootstrapper.a, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
  ld: warning: ignoring file /Users/chr/.nuget/packages/runtime.osx-arm64.microsoft.dotnet.ilcompiler/8.0.0-alpha.1.22559.2/sdk/libstdc++compat.a, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
  ld: warning: ignoring file obj/Debug/net8.0/osx-arm64/native/helloworldapp.o, building for macOS-x86_64 but attempting to link with file built for unknown-arm64
  ld: warning: ignoring file /Users/chr/.nuget/packages/runtime.osx-arm64.microsoft.dotnet.ilcompiler/8.0.0-alpha.1.22559.2/framework/libSystem.Security.Cryptography.Native.Apple.a, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
  ld: warning: ignoring file /Users/chr/.nuget/packages/runtime.osx-arm64.microsoft.dotnet.ilcompiler/8.0.0-alpha.1.22559.2/sdk/libRuntime.WorkstationGC.a, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
  ld: warning: ignoring file /Users/chr/.nuget/packages/runtime.osx-arm64.microsoft.dotnet.ilcompiler/8.0.0-alpha.1.22559.2/framework/libSystem.Security.Cryptography.Native.OpenSsl.a, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
  ld: warning: ignoring file /Users/chr/.nuget/packages/runtime.osx-arm64.microsoft.dotnet.ilcompiler/8.0.0-alpha.1.22559.2/framework/libSystem.Globalization.Native.a, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
  ld: warning: ignoring file /Users/chr/.nuget/packages/runtime.osx-arm64.microsoft.dotnet.ilcompiler/8.0.0-alpha.1.22559.2/framework/libSystem.Native.a, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
  ld: warning: ignoring file /Users/chr/.nuget/packages/runtime.osx-arm64.microsoft.dotnet.ilcompiler/8.0.0-alpha.1.22559.2/framework/libSystem.IO.Compression.Native.a, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
  Undefined symbols for architecture x86_64:
    "_main", referenced from:
       implicit entry/start for main executable
  ld: symbol(s) not found for architecture x86_64
  clang: error: linker command failed with exit code 1 (use -v to see invocation)
/Users/chr/.nuget/packages/microsoft.dotnet.ilcompiler/8.0.0-alpha.1.22559.2/build/Microsoft.NETCore.Native.targets(346,5): error MSB3073: The command ""clang" "obj/Debug/net8.0/osx-arm64/native/helloworldapp.o" -o "bin/Debug/net8.0/osx-arm64/native/helloworldapp" /Users/chr/.nuget/packages/runtime.osx-arm64.microsoft.dotnet.ilcompiler/8.0.0-alpha.1.22559.2/sdk/libbootstrapper.a /Users/chr/.nuget/packages/runtime.osx-arm64.microsoft.dotnet.ilcompiler/8.0.0-alpha.1.22559.2/sdk/libRuntime.WorkstationGC.a /Users/chr/.nuget/packages/runtime.osx-arm64.microsoft.dotnet.ilcompiler/8.0.0-alpha.1.22559.2/sdk/libstdc++compat.a /Users/chr/.nuget/packages/runtime.osx-arm64.microsoft.dotnet.ilcompiler/8.0.0-alpha.1.22559.2/framework/libSystem.Native.a /Users/chr/.nuget/packages/runtime.osx-arm64.microsoft.dotnet.ilcompiler/8.0.0-alpha.1.22559.2/framework/libSystem.Globalization.Native.a /Users/chr/.nuget/packages/runtime.osx-arm64.microsoft.dotnet.ilcompiler/8.0.0-alpha.1.22559.2/framework/libSystem.IO.Compression.Native.a /Users/chr/.nuget/packages/runtime.osx-arm64.microsoft.dotnet.ilcompiler/8.0.0-alpha.1.22559.2/framework/libSystem.Net.Security.Native.a /Users/chr/.nuget/packages/runtime.osx-arm64.microsoft.dotnet.ilcompiler/8.0.0-alpha.1.22559.2/framework/libSystem.Security.Cryptography.Native.Apple.a /Users/chr/.nuget/packages/runtime.osx-arm64.microsoft.dotnet.ilcompiler/8.0.0-alpha.1.22559.2/framework/libSystem.Security.Cryptography.Native.OpenSsl.a -g -Wl,-rpath,'@executable_path' -ldl -lm -lobjc -lswiftCore -lswiftFoundation -lz -licucore -L/usr/lib/swift -framework CoreFoundation -framework CryptoKit -framework Foundation -framework Security -framework GSS" exited with code 1. [/Users/chr/d/d2/helloworldapp/helloworldapp.csproj]
```

</details>

--- a/src/coreclr/nativeaot/BuildIntegration/Microsoft.DotNet.ILCompiler.SingleEntry.targets
+++ b/src/coreclr/nativeaot/BuildIntegration/Microsoft.DotNet.ILCompiler.SingleEntry.targets
@@ -17,9 +17,8 @@
 
     <RuntimeIlcPackageName>runtime.$(OSIdentifier)-$(TargetArchitecture).Microsoft.DotNet.ILCompiler</RuntimeIlcPackageName>
 
-    <OSHostArch>$([System.Runtime.InteropServices.RuntimeInformation]::OSArchitecture.ToString().ToLowerInvariant)</OSHostArch>
-
-    <IlcHostArch Condition="'$(IlcHostArch)' == ''">$(OSHostArch)</IlcHostArch>
+    <!-- Default to host that matches SDK architecture -->
+    <IlcHostArch Condition="'$(IlcHostArch)' == ''">$([System.Runtime.InteropServices.RuntimeInformation]::ProcessArchitecture.ToString().ToLowerInvariant)</IlcHostArch>
     <IlcHostPackageName>runtime.$(OSIdentifier)-$(IlcHostArch).Microsoft.DotNet.ILCompiler</IlcHostPackageName>
     
     <IlcCalledViaPackage>true</IlcCalledViaPackage>

